uat/src/main/java/com/aws/greengrass/testing/mqtt5/client/grpc/GRPCControlServer.java

```diff
@@ -273,33 +288,217 @@ public void publishMqtt(MqttPublishRequest request, StreamObserver<MqttPublishRe
+                                                .asRuntimeException());
+                    return;
+                }
+
+                int retainHandling = subscription.getRetainHandlingValue();
+                if (retainHandling < RETAIN_HANDLING_MIN || retainHandling > RETAIN_HANDLING_MAX) {
+                    logger.atWarn().log("qos is invalid can be in range [0,3] but is {} at index {}", qos, index);
+                    responseObserver.onError(Status.INVALID_ARGUMENT
+                                                .withDescription("qos is invalid can be in range [0,3]")
+                                                .asRuntimeException());
+                    return;
+                }
+
+                boolean noLocal = subscription.getNoLocal();
+                boolean retainAsLocal = subscription.getRetainAsPublished();
+                MqttConnection.Subscription tmp = new MqttConnection.Subscription(filter, qos, noLocal, retainAsLocal,
+                                                                                    retainHandling);
+                logger.atInfo().log("Subscription: filter {} QoS {} noLocal {} retainAsLocal {} retainHandling {}",
+                                        filter, qos, noLocal, retainAsLocal, retainHandling);
+                outSubscriptions.add(tmp);
+                index++;
+            }
+
+            int connectionId = request.getConnectionId().getConnectionId();
+            MqttConnection connection = mqttLib.getConnection(connectionId);
+            if (connection == null) {
+                logger.atWarn().log(CONNECTION_WITH_DOES_NOT_FOUND, connectionId);
+                responseObserver.onError(Status.NOT_FOUND
+                                            .withDescription(CONNECTION_DOES_NOT_FOUND)
+                                            .asRuntimeException());
+                return;
+            }
+
+            logger.atInfo().log("Subscribe: connectionId {} subscriptionId {} for {} filters",
+                                    connectionId, subscriptionId, outSubscriptions.size());
+            MqttSubscribeReply.Builder builder = MqttSubscribeReply.newBuilder();
+            try {
+                // TODO: pass also user's properties
+                MqttConnection.SubAckInfo subAckInfo = connection.subscribe(timeout, subscriptionId, outSubscriptions);
+                if (subAckInfo != null) {
+                    List<Integer> codes = subAckInfo.getReasonCodes();
+                    if (codes != null) {
+                        builder.addAllReasonCodes(codes);
+                    }
+
+                    String reasonString = subAckInfo.getReasonString();
+                    if (reasonString != null) {
+                         builder.setReasonString(reasonString);
+                    }
+                    logger.atInfo().log("Subscribe response: reason codes {} reason string {}", codes, reasonString);
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 429:440, 491:502.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.